From c10c0ad125a4f9f330e6bb5612b08b6010e851af Mon Sep 17 00:00:00 2001
From: Valentine Barshak <vbarshak@dev.rtsoft.ru>
Date: Wed, 10 Apr 2013 16:51:17 +0400
Subject: [PATCH 053/127] arm: tegra: pcie: Use PM notifier instead of device
 PM callbacks

Some devices may need to load firmware when PCIe controller wakes
up and re-scans its busses. Firmware operations will fail when done
from the suspend/resume callbacks because usermode helpers are
disabled at this point.

Switch to using PM notifier callbacks instead. The PM_SUSPEND_PREPARE
notification is done early before the drivers are asked to go to sleep.
The PM_POST_SUSPEND is sent when the system has just resumed and
the usermodehelper is enabled.

Signed-off-by: Valentine Barshak <vbarshak@dev.rtsoft.ru>
---
 arch/arm/mach-tegra/pcie.c |   56 +++++++++++++++++++++++++++++++------------
 1 files changed, 40 insertions(+), 16 deletions(-)

diff --git a/arch/arm/mach-tegra/pcie.c b/arch/arm/mach-tegra/pcie.c
index bac6eee..7f9d70f 100644
--- a/arch/arm/mach-tegra/pcie.c
+++ b/arch/arm/mach-tegra/pcie.c
@@ -37,6 +37,7 @@
 #include <linux/platform_device.h>
 #include <linux/regulator/consumer.h>
 #include <linux/workqueue.h>
+#include <linux/suspend.h>
 #include <linux/gpio.h>
 
 #include <asm/sizes.h>
@@ -324,6 +325,9 @@ struct tegra_pcie_info {
 	int			power_rails_enabled;
 	int			pcie_power_enabled;
 	struct work_struct 	hotplug_detect;
+#ifdef CONFIG_PM
+	struct notifier_block	pm_nb;
+#endif
 
 	struct regulator	*regulator_hvdd;
 	struct regulator	*regulator_pexio;
@@ -614,8 +618,10 @@ static struct hw_pci tegra_pcie_hw = {
 };
 
 #ifdef CONFIG_PM
-static int tegra_pcie_suspend(struct device *dev);
-static int tegra_pcie_resume(struct device *dev);
+static int tegra_pcie_suspend(void);
+static int tegra_pcie_resume(void);
+static int tegra_pcie_pm_handler(struct notifier_block *nb,
+					unsigned long val, void *ign);
 
 /* It enumerates the devices when dock is connected after system boot */
 /* this is similar to pcibios_init_hw in bios32.c */
@@ -653,14 +659,18 @@ static void tegra_pcie_hotplug_init(void)
 static void tegra_pcie_attach(void)
 {
 #ifdef CONFIG_PM
-	tegra_pcie_resume(NULL);
+	mutex_lock(&pm_mutex);
+	tegra_pcie_resume();
+	mutex_unlock(&pm_mutex);
 #endif
 }
 
 static void tegra_pcie_detach(void)
 {
 #ifdef CONFIG_PM
-	tegra_pcie_suspend(NULL);
+	mutex_lock(&pm_mutex);
+	tegra_pcie_suspend();
+	mutex_unlock(&pm_mutex);
 #endif
 }
 
@@ -1164,6 +1174,13 @@ static int tegra_pcie_get_resources(void)
 		goto err_pwr_on;
 	}
 	set_irq_flags(INT_PCIE_INTR, IRQF_VALID);
+
+#ifdef CONFIG_PM
+	tegra_pcie.pm_nb.notifier_call = tegra_pcie_pm_handler;
+	tegra_pcie.pm_nb.priority = 0;
+	register_pm_notifier(&tegra_pcie.pm_nb);
+#endif
+
 	return 0;
 
 err_pwr_on:
@@ -1371,7 +1388,7 @@ static int tegra_pcie_probe(struct platform_device *pdev)
 }
 
 #ifdef CONFIG_PM
-static int tegra_pcie_suspend(struct device *dev)
+static int tegra_pcie_suspend(void)
 {
 	struct pci_dev *pdev;
 
@@ -1404,7 +1421,7 @@ static void tegra_pcie_set_irq(struct pci_bus *bus)
 	}
 }
 
-static int tegra_pcie_resume(struct device *dev)
+static int tegra_pcie_resume(void)
 {
 	int ret = 0;
 	struct pci_bus *bus = NULL;
@@ -1461,6 +1478,23 @@ static int tegra_pcie_resume(struct device *dev)
 exit:
 	return 0;
 }
+
+static int tegra_pcie_pm_handler(struct notifier_block *nb,
+					unsigned long val, void *ign)
+{
+	switch (val) {
+	case PM_POST_SUSPEND:
+		tegra_pcie_resume();
+		break;
+	case PM_SUSPEND_PREPARE:
+		tegra_pcie_suspend();
+		break;
+	default:
+		return NOTIFY_DONE;
+	}
+
+	return NOTIFY_OK;
+}
 #endif
 
 static int tegra_pcie_remove(struct platform_device *pdev)
@@ -1468,22 +1502,12 @@ static int tegra_pcie_remove(struct platform_device *pdev)
 	return 0;
 }
 
-#ifdef CONFIG_PM
-static const struct dev_pm_ops tegra_pcie_pm_ops = {
-	.suspend = tegra_pcie_suspend,
-	.resume = tegra_pcie_resume,
-	};
-#endif
-
 static struct platform_driver tegra_pcie_driver = {
 	.probe   = tegra_pcie_probe,
 	.remove  = tegra_pcie_remove,
 	.driver  = {
 		.name  = "tegra-pcie",
 		.owner = THIS_MODULE,
-#ifdef CONFIG_PM
-		.pm    = &tegra_pcie_pm_ops,
-#endif
 	},
 };
 
-- 
1.7.4.4

